File: sorts/relative_sort_array.rb
def relative_sort_array(arr1, arr2)
    return arr1 if arr1.empty?

    freq = Hash.new(0)
    arr1.each { |num| freq[num] += 1}

    result = []

    arr2.each do |num|
        while freq[num] > 0
            result << num
            freq[num] -= 1
        end
    end

    freq.keys.sort.each do |num| 
        freq[num].times {result << num }
    end

    result
end

File: sorts/relative_sort_array.py
class Solution:
    def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
      if not arr1:
        return arr1

      freq = {}
      for num in arr1:
        freq[num] = freq.get(num, 0) + 1

      result = []

      for num in arr2:
        while freq.get(num, 0) > 0:
          result.append(num)
          freq[num] -= 1
        
        if num in freq:
            del freq[num]

      
      for num in sorted(freq.keys()):
        result.extend([num] * freq[num])

      return result